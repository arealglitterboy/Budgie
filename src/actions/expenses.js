import {v4 as uuid} from 'uuid';

export const addExpense = (
    {
        description = '',
        note = '',
        date = 0,
        currency = 'â‚¬',
        amount = 0
    } = {}
    ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        date,
        currency,
        amount
    }
});

export const removeExpense = ({ id }) => ({ type: 'REMOVE_EXPENSE', id });

export const editExpense = (id, updates) => ({ type: 'EDIT_EXPENSE', id, updates });