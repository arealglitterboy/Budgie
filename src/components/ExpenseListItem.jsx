import React from 'react';
import { Link } from 'react-router-dom';

const times = {
    year:   31536000000, //(1000 * 60 * 60 * 24 * 365)
    month:  2419200000,  //(1000 * 60 * 60 * 24 * 7 * 4)
    week:   604800000,   //(1000 * 60 * 60 * 24 * 7)
    day:    86400000,    //(1000 * 60 * 60 * 24)
    // hour:   3600000,     //(1000 * 60 * 60)
    // minute: 60000,       //(1000 * 60)
    // second: 1000,        //(1000)
};

const formatDifference = (time, unit, inPast) => (((!inPast) ? "In " : "") + time + " " + unit + ((time != 1) ? 's' : '') + ((inPast) ? " ago" : ""));

function findDifference(date) {
    let today = Date.now();
    const difference = Math.abs(date - today);
    const key = Object.keys(times).find((time) => (difference >= times[time]));
    return (key) ? (formatDifference(Math.floor(difference/times[key]), key, (date < today))) : "Today";
}

export default class ExpenseListItem extends React.Component {
    // Open up the drop down portion with note and edit button
    // ? Might be easier to use if, when the search is coming up positive due to the string being found in the note, open the note?
    handleClick = (e) => {
        console.log(this);
        console.log(e);
    };
    
    render() {
        const { description, note, date, amount, currency, id} = this.props;
        const time = new Date(date);
        
        return (
            <article className="expense-item" onClick={this.handleClick}>
                <img src="" alt="icon" className="expense-item__icon" />
                
                <section className="expense-item__info">
                    <time className="expense-item__info__date" dateTime={`${time.toUTCString()}`}>{findDifference(date)}</time>
                    <h5 className="expense-item__info__description">{description}</h5> 
                </section>
                
                <data className="expense-item__amount" value={amount}>
                    {(amount/100).toLocaleString('en-US', { style: "currency", currency })}
                </data>
                
                <footer className="expense-item__footer expense-item__footer--inactive">
                    {(note) && <p>{note}</p>}
                    <Link to={`/edit/${id}`} className="expense-item__footer__edit">EDIT</Link>
                </footer>
            </article>
        );
    }
}