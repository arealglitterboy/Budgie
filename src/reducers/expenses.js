const defaultState = [];

export default (state = defaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((e) => (e.id !== action.id));
        case 'EDIT_EXPENSE': // ? Map each value in the array into a new array, if the id matches the selected id map it with its new values
            return state.map((e) => ((e.id === action.id) ? {...e, ...action.updates} : {...e}));
        default:
            return state;
    }
};