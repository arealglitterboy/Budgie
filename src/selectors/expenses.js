import moment from "moment";

const includesIgnoreCase = (key, searching) => (!key || (searching && searching.toLowerCase().includes(key.toLowerCase())));

/**
 * 
 * @param {string} search 
 * @returns {function}
 */
const createIncludes = (search) => {
    const regExp = new RegExp(search, 'gi')
    return (str) => str.search(regExp) >= 0;
};

function filterExpense({ description, note, date }, term, startDate, endDate) {
    const includes = createIncludes(term);
    const dateMoment = moment(date);

    const containsTerm = includes(description) || includes(note); // ? If either the description or the note contains the search term.
    // const containsTerm = includesIgnoreCase(term, description) || includesIgnoreCase(term, note); // ? If either the description or the note contains the search term.
    const inDate = (!startDate || startDate.isSameOrBefore(dateMoment, 'day')) && (!endDate || endDate.isSameOrAfter(dateMoment, 'day')); // ? If the start/end date is defined, and the given date is greater/less than it.
    return containsTerm && inDate;
}

export default (expenses, { term, startDate, endDate, sortFunction }) => {
    return expenses.filter((e) => (filterExpense(e, term, startDate, endDate))).sort(sortFunction);
};