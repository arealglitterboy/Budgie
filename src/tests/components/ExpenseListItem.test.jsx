import React from "react";
import { shallow } from "enzyme";

import { expenses } from "../fixtures/expenses.fixture";
import ExpenseListItem from "../../components/ExpenseListItem";

test('should render Expense List Item with expense details', () => {
    const wrapper = shallow(<ExpenseListItem { ...expenses[0] } today={new Date('2021-09-12')} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Expense List Item with expense details and different category', () => {
    const wrapper = shallow(<ExpenseListItem { ...expenses[2] } today={new Date('2021-09-12')} />)
    expect(wrapper).toMatchSnapshot();
});