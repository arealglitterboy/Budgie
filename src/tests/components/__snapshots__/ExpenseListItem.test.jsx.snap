// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List Item with expense details 1`] = `
<article
  className="expense"
  onClick={[Function]}
>
  <header
    className="expense__header"
  >
    <time
      className="expense__header__date"
      dateTime="Fri, 09 Apr 2021 00:00:00 GMT"
    >
      5 months ago
    </time>
  </header>
  <section
    className="expense__body"
  >
    <img
      alt="icon"
      className="expense__body__icon"
      src=""
    />
    <h5
      className="expense__body__title"
    >
      Food
    </h5>
    <h6
      className="expense__body__name"
    />
    <data
      className="expense__amount"
      value={1050}
    >
      €10.50
    </data>
  </section>
  <footer
    className="expense__footer expense__footer--inactive"
  >
    <p>
      Bought food for mastication
    </p>
    <Link
      className="expense__footer__edit"
      to="/edit/1"
    >
      EDIT
    </Link>
  </footer>
</article>
`;

exports[`should render Expense List Item with expense details and different categories 1`] = `
<article
  className="expense"
  onClick={[Function]}
>
  <header
    className="expense__header"
  >
    <time
      className="expense__header__date"
      dateTime="Fri, 16 Jul 2010 00:00:00 GMT"
    >
      11 years ago
    </time>
  </header>
  <section
    className="expense__body"
  >
    <img
      alt="icon"
      className="expense__body__icon"
      src=""
    />
    <h5
      className="expense__body__title"
    >
      Credit Card
    </h5>
    <h6
      className="expense__body__name"
    />
    <data
      className="expense__amount"
      value={4500}
    >
      €45.00
    </data>
  </section>
  <footer
    className="expense__footer expense__footer--inactive"
  >
    <p>
      Paying off loan to completion
    </p>
    <Link
      className="expense__footer__edit"
      to="/edit/3"
    >
      EDIT
    </Link>
  </footer>
</article>
`;
